src/crowdin_client.cpp

```diff
@@ -135,150 +159,332 @@ CrowdinClient::~CrowdinClient() {}
-                                                 const Language& lang,
+dispatch::future<void> CrowdinClient::UploadFile(const long project_id,
+                                                 const long file_id,
+                                                 const std::string& lang_tag,
                                                  const std::string& file_content)
 {
-    auto url = "/api/project/" + project_id + "/upload-translation";
-
-    multipart_form_data data;
-    data.add_value("json", "");
-    data.add_value("language", lang.LanguageTag());
-    data.add_value("import_duplicates", "0");
-    data.add_value("import_eq_suggestions", "0");
-    data.add_file("files[" + str::to_utf8(file) + "]", "upload.po", file_content);
-
-    return m_api->post(url, data).then([](json){});
+    return RefreshToken().then([=] () {
+    return m_api->post(
+            "storages",
+            octet_stream_data(file_content),
+            { { "Crowdin-API-FileName", L"poedit.xliff"} }
+        )
+        .then([this, project_id, file_id, lang_tag] (json r) {
+            cout << "File uploaded to temporary storage: " << r << "\n\n";
+            const auto storageId = r["data"]["id"].get<int>();
+            return m_api->post(
+                "projects/" + std::to_string(project_id) + "/translations/" + lang_tag,
+                json_data(json({
+                    { "storageId", storageId },
+                    { "fileId", file_id },
+                    { "importDuplicates", true },
+                    { "autoApproveImported", true }
+                })))
+                .then([](json r) {
+                    cout << "File uploaded: " << r << "\n\n";
+                });
+        });
+    });
 }
 
-
-bool CrowdinClient::IsSignedIn() const
+json CrowdinClient::LoadAuth()
 {
-    std::string token;
-    return keytar::GetPassword("Crowdin", "", &token);
+    std::string auth;
+    return keytar::GetPassword("Crowdin", "", &auth)
+            // if not came from older (<2.4) version
+            && auth[0] == '{'
+                ? json::parse(auth)
+                : json();
 }
 
-
-void CrowdinClient::SignInIfAuthorized()
+void CrowdinClient::SetAuth(const json& auth)
 {
-    std::string token;
-    if (keytar::GetPassword("Crowdin", "", &token))
-        SetToken(token);
-}
+    cout << endl << endl << "Authorization: " << auth << endl << endl;
+
+    if(auth.empty()) {
+        return;
+    }
 
+    std::string access_token = auth["access_token"];
+    m_authRefreshToken = auth["refresh_token"].get<string>();
+    m_authExpireTime = auth["expiration_time"].get<time_t>();
+    
+    auto domain = jwt::decode(access_token).get_payload_claim("domain");
+    if(domain.get_type() == jwt::claim::type::null) {
+        m_api = make_unique<crowdin_http_client>(*this, "https://crowdin.com/api/v2");
+    } else {
+        m_api = make_unique<crowdin_http_client>(*this, "https://" + domain.as_string() + ".crowdin.com/api/v2");
```

Comment:
This requirement that http_client only works with URLs with the same prefix is quite annoying and causing you no end of trouble, isn't it. It comes from C++ REST SDK used on non-macOS and it seems it is deeply routed there. I'm trying to figure out if there's some way around it to make our lives easier.

Reply:
Marking as resolved unless you have any additional comments/requirements to it, or please feel free to comment/unresolve if otherwise